src/Orleans.Runtime/GrainDirectory/GrainDirectoryHandoffManager.cs

```diff
@@ -212,37 +208,79 @@ internal void ProcessSiloAddEvent(DirectoryMembershipSnapshot membershipSnapshot
+                    // adjust partitions by splitting them accordingly between new and old silos
+                    SiloAddress predecessorOfNewSilo = membershipSnapshot.FindPredecessor(addedSilo);
+                    if (!directoryPartitionsMap.TryGetValue(predecessorOfNewSilo, out var partitionToSplit))
+                    {
+                        // we should have the partition of the predcessor of our new successor
+                        logger.LogWarning(
+                            (int)ErrorCode.DirectoryPartitionPredecessorExpected,
+                            "This silo is expected to hold directory partition of {Predecessor}",
+                            predecessorOfNewSilo);
+                    }
+                    else
                     {
-                        var s = membershipSnapshot.CalculateGrainDirectoryPartition(grain);
-                        return (s != null) && !this.localSiloDetails.SiloAddress.Equals(s);
-                    }, false);
-                List<ActivationAddress> splitPartListSingle = splitPart.ToListOfActivations(true);
-                List<ActivationAddress> splitPartListMulti = splitPart.ToListOfActivations(false);
-
-                EnqueueOperation(
-                    $"{nameof(ProcessSiloAddEvent)}({addedSilo})",
-                    () => ProcessAddedSiloAsync(membershipSnapshot, addedSilo, splitPartListSingle, splitPartListMulti));
-            
+                        if (logger.IsEnabled(LogLevel.Debug))
+                        {
+                            logger.LogDebug("Splitting partition of {Predecessor} and creating a copy for {AddedSilo}", predecessorOfNewSilo, addedSilo);
+                        }
+
+                        GrainDirectoryPartition splitPart = partitionToSplit.Split(
+                            grain =>
+                            {
+                                // Need to review the 2nd line condition.
+                                var s = membershipSnapshot.CalculateGrainDirectoryPartition(grain);
+                                return (s != null) && !predecessorOfNewSilo.Equals(s);
+                            }, true);
+                        directoryPartitionsMap[addedSilo] = splitPart;
+                    }
+                }
+
                 // remove partition of one of the old successors that we do not need to now
                 SiloAddress oldSuccessor = directoryPartitionsMap.FirstOrDefault(pair => !successor.Equals(pair.Key)).Key;
                 if (oldSuccessor == null) return;
 
-                if (logger.IsEnabled(LogLevel.Debug)) logger.Debug("Removing copy of the directory partition of silo " + oldSuccessor + " (holding copy of " + addedSilo + " instead)");
+                if (logger.IsEnabled(LogLevel.Debug))
+                {
+                    logger.LogDebug(
+                        "Removing copy of the directory partition of silo {OldSuccessor} (holding copy of {AddedSilo} instead)",
```

Comment:
A tangential note, I wonder at which point and which way, if at all, it would make sense to move into using https://docs.microsoft.com/fi-fi/dotnet/api/microsoft.extensions.logging.eventid ? As one big refactoring or a plugin at a time? I could imagine this could be tied to the refactoring @xiazen is doing about loggin and defining a set of IDs (as `const` with appropriate messages) and then mandating/testing any one is only used once. Would make add to choices to build analytics on running clusters (SIEM). An example I have in mind (just composing something here as an example):

```
/// <summary>
    /// Ties together the event ID and corresponding format string.
    /// </summary>
    [DebuggerDisplay("EventStruct(Id = Id, FormatString = {FormatString})")]
    public readonly struct EventStruct
    {
        /// <summary>
        /// The ID of the logging event. This should be unique per purpose.
        /// </summary>
        public EventId Id { get; }

        /// <summary>
        /// The string used to format the message related to the <see cref="Id"/>.
        /// </summary>
        public string FormatString { get; }


        /// <summary>
        /// Constructs an event together with its formatting string.
        /// </summary>
        /// <param name="id">The ID of the logging event. This should be unique per purpose.</param>
        /// <param name="formatString">The string used to format the message related to the <paramref name="id"/>.</param>
        public EventStruct(in EventId id, string formatString)
        {
            Id = id;
            FormatString = formatString ?? throw new ArgumentNullException(nameof(formatString));
        }
    }

 /// <summary>
    /// Events used to log application activities.
    /// </summary>
    public static class Events
    {
        /// <summary>
        /// Something, something...
        /// </summary>
       /// <remarks>Check called only once in the framework.</remarks>
        public static EventStruct Something { get; } = new EventStruct(new EventId(1, nameof(Something)), "Something in {something}.");
    }
```

Reply:
Leveraging `EventId` and the [`LoggerMessage.Define` pattern](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/loggermessage?view=aspnetcore-2.2) would be beneficial, but I want to save it for another time